packages/block-editor/src/hooks/text-align.js

```diff
@@ -0,0 +1,198 @@
+	if ( 'type' in ( settings.attributes?.textAlign ?? {} ) ) {
+		return settings;
+	}
+	if ( hasBlockSupport( settings, 'textAlign' ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			textAlign: {
+				type: 'string',
+				// Allow for '' since it is used by the `updateTextAlignment` function
+				// in toolbar controls for special cases with defined default values.
+				enum: [ ...VALID_TEXT_ALIGNMENTS, '' ],
+			},
+		};
+	}
+
+	return settings;
+}
+
+function BlockEditTextAlignmentToolbarControlsPure( {
+	name: blockName,
+	textAlign,
+	setAttributes,
+} ) {
+	const validTextAlignments = getValidTextAlignments(
+		getBlockSupport( blockName, 'textAlign' )
+	);
+	const blockEditingMode = useBlockEditingMode();
+	if ( ! validTextAlignments.length || blockEditingMode !== 'default' ) {
+		return null;
+	}
+
+	const textAlignmentControls = TEXT_ALIGNMENT_OPTIONS.filter( ( control ) =>
+		validTextAlignments.includes( control.align )
+	);
+
+	const updateTextAlignment = ( nextTextAlign ) => {
+		if ( ! nextTextAlign ) {
+			const blockType = getBlockType( blockName );
+			const blockDefaultTextAlign =
+				blockType?.attributes?.textAlign?.default;
+			if ( blockDefaultTextAlign ) {
+				nextTextAlign = '';
+			}
+		}
+		setAttributes( { textAlign: nextTextAlign } );
+	};
+
+	return (
+		<BlockControls group="block">
+			<AlignmentControl
+				value={ textAlign }
+				onChange={ updateTextAlignment }
+				alignmentControls={ textAlignmentControls }
+			/>
+		</BlockControls>
+	);
+}
+
+export default {
+	edit: BlockEditTextAlignmentToolbarControlsPure,
+	useBlockProps,
+	addSaveProps: addAssignedTextAlign,
+	attributeKeys: [ 'textAlign' ],
+	hasSupport( name ) {
+		return hasBlockSupport( name, 'textAlign', false );
+	},
+};
+
+function useBlockProps( { name, textAlign } ) {
+	const validTextAlignments = getValidTextAlignments(
+		getBlockSupport( name, 'textAlign' )
+	);
+	if ( ! validTextAlignments.length ) {
+		return null;
+	}
+	const className = classnames( {
+		[ `has-text-align-${ textAlign }` ]: textAlign,
+	} );
+	return { className };
+}
+
+/**
+ * Override props assigned to save component to inject text alignment class
+ * name if block supports it.
+ *
+ * @param {Object} props      Additional props applied to save element.
+ * @param {Object} blockType  Block type.
+ * @param {Object} attributes Block attributes.
+ *
+ * @return {Object} Filtered props applied to save element.
+ */
+export function addAssignedTextAlign( props, blockType, attributes ) {
+	const { textAlign } = attributes;
+	const blockTextAlign = getBlockSupport( blockType, 'textAlign' );
+	const isTextAlignValid =
+		getValidTextAlignments( blockTextAlign ).includes( textAlign );
+	if ( isTextAlignValid ) {
+		props.className = classnames(
+			`has-text-align-${ textAlign }`,
+			props.className
+		);
+	}
+	return props;
+}
+
+addFilter(
+	'blocks.registerBlockType',
```

Comment:
It would be cool to abstract this hook into the default export (an addAttribute property or something) but I guess that's separate from this PR.